asmcomp/arm64/emit.mlp

```diff
@@ -767,34 +762,38 @@ let emit_instr env i =
               reg_tmp1 in
         begin match memory_chunk with
         | Byte_unsigned ->
-            `	{emit_string load_type}b	{emit_wreg dst}, {emit_addressing addressing_mode base}\n`
+            `	ldrb	{emit_wreg dst}, {emit_addressing addressing_mode base}\n`
         | Byte_signed ->
-            `	{emit_string load_type}sb	{emit_reg dst}, {emit_addressing addressing_mode base}\n`
+            `	ldrsb	{emit_reg dst}, {emit_addressing addressing_mode base}\n`
         | Sixteen_unsigned ->
-            `	{emit_string load_type}h	{emit_wreg dst}, {emit_addressing addressing_mode base}\n`
+            `	ldrh	{emit_wreg dst}, {emit_addressing addressing_mode base}\n`
         | Sixteen_signed ->
-            `	{emit_string load_type}sh	{emit_reg dst}, {emit_addressing addressing_mode base}\n`
+            `	ldrsh	{emit_reg dst}, {emit_addressing addressing_mode base}\n`
         | Thirtytwo_unsigned ->
-            `	{emit_string load_type}	{emit_wreg dst}, {emit_addressing addressing_mode base}\n`
+            `	ldr	{emit_wreg dst}, {emit_addressing addressing_mode base}\n`
         | Thirtytwo_signed ->
-            `	{emit_string load_type}sw	{emit_reg dst}, {emit_addressing addressing_mode base}\n`
+            `	ldrsw	{emit_reg dst}, {emit_addressing addressing_mode base}\n`
         | Single ->
-            `	{emit_string load_type}	s7, {emit_addressing addressing_mode base}\n`;
+            `	ldr	s7, {emit_addressing addressing_mode base}\n`;
             `	fcvt	{emit_reg dst}, s7\n`
-        | Word_int | Word_val | Double ->
-            `	{emit_string load_type}	{emit_reg dst}, {emit_addressing addressing_mode base}\n`
+        | Word_int | Word_val ->
+            if is_atomic then begin
+              assert (addressing_mode = Iindexed 0);
+              `	ldar	{emit_reg dst}, [{emit_reg i.arg.(0)}]\n`
+            end else
+              `	ldr	{emit_reg dst}, {emit_addressing addressing_mode base}\n`
+        | Double ->
+            `	ldr	{emit_reg dst}, {emit_addressing addressing_mode base}\n`
         end
-    | Lop(Istore(size, addr, initialization)) ->
-        (* memory model barrier for non-initializing store *)
-        if not initialization then ` dmb ishld\n`;
+    | Lop(Istore(size, addr, assignment)) ->
         let src = i.arg.(0) in
```

Comment:
The code (I think) is assuming that `size = Word_int || size = Word_val || assignment = false`. 
```suggestion
        assert(size = Word_int || size = Word_val || assignment = false);
        let src = i.arg.(0) in
```

Reply:
Noooo!  There are plenty of assignments at other sizes, e.g. in byte arrays, flat float arrays, bigarrays.  But we agreed (during the memory model meeting) that these assignments will not follow the Multicore OCaml memory model, so no barrier is needed.

Reply:
OK for the comment.  Memory model discussions tend to persist and repeat whatever we do :-),  but the comment is a start.